### PR TITLE
Better query params and fix #69

### LINKED ISSUE #69: [Question/BUG] Query Parameters converted to float
# issue / feature request 

I Would like to understand the rationale of converting any query parameter that is a series of digits to floats. I am building an API that relies on Account ID's being based which are digits, these are needed in string form and python-lambdarest uses the `float()` function to convert all strings that are numeric, without the ability to override this behaviour from what i can tell. 

Is there a reason for this - HTTP is a string based protocol afaik, hence converting to float seems somewhat arbitrary and if needed something that could be done in business logic vs the underlying library which is parsing the API Gateway JSON payload.

Thanks!

### PR DESCRIPTION
**Pull request by TPRobots**
Pull request opened by github-pullrequestcreator.

### CODE REVIEW

File: lambdarest/__init__.py [hunk 1]
```diff
@@ -74,28 +74,74 @@ class ScopeMissing(Exception):
...
+    def inner_cast(inner_type):
+        try:
+            return list(map(inner_type, values_list))
+        except TypeError:
+            pass
+        except ValueError:
+            pass
+        return None
+
+    return inner_cast(type) or inner_cast(str) or value
```
Comment (tmo-trustpilot) [inline lambdarest/__init__.py:89]: This is cool 😄

File: lambdarest/__init__.py [hunk 2]
```diff
@@ -74,28 +74,74 @@ class ScopeMissing(Exception):
...
+def __json_load_query(query, query_param_schema=None):
     query = query or {}
+    query_param_schema = query_param_schema or {}
 
-    return {key: __marshall_query_params(value) for key, value in query.items()}
+    return {
+        key: __marshall_value(value, query_param_schema.get(key, {}))
+        for key, value in query.items()
+    }
+    print(r)
```
Comment (tmo-trustpilot) [inline lambdarest/__init__.py:143]:
```suggestion
```

File: docs/README.md [hunk 3]
```diff
@@ -114,7 +114,11 @@ assert result == {"body": 'Validation Error', "statusCode": 400, "headers":{}}
 ## Query Params
 
 Query parameters are also analyzed and validatable with JSON schemas.
-Query arrays are expected to be comma separated, all numbers are converted to floats.
+
+Query arrays are expected to be comma separated.
+
+All values are unpacked to types defined in `schema.properties.query.properties.*`.
+
```
Comment (tmo-trustpilot) [inline docs/README.md:121]: Could you put a simple example here?

File: docs/README.md [hunk 4]
```diff
@@ -156,6 +160,14 @@ result = lambda_handler(event=valid_input_event)
 assert result == {"body": '{"foo": [1.0, 2.2, 3.0]}', "statusCode": 200, "headers":{}}
 ```
 
+### arrays without specification in json schema
```
Comment (tmo-trustpilot) [inline docs/README.md:163]:
```suggestion
### Arrays without specification in json schema
```

File: docs/README.md [hunk 5]
```diff
@@ -156,6 +160,14 @@ result = lambda_handler(event=valid_input_event)
...
+
+If you supply an array, (comma seperated value),  to a query arg and you *dont* include a specification for the field in your jsonschema *then*:
```
Comment (tmo-trustpilot) [inline docs/README.md:165]:
```suggestion
If you supply a comma separated value to a query arg and you *don't* include a specification for the field in your jsonschema *then*: 
```

File: docs/README.md [hunk 6]
```diff
@@ -156,6 +160,14 @@ result = lambda_handler(event=valid_input_event)
 assert result == {"body": '{"foo": [1.0, 2.2, 3.0]}', "statusCode": 200, "headers":{}}
 ```
 
+### arrays without specification in json schema
+
+If you supply an array, (comma seperated value),  to a query arg and you *dont* include a specification for the field in your jsonschema *then*: 
+
+1. the value will be unpacked as a string array
+2. all values will be tried cast to floats
```
Comment (tmo-trustpilot) [inline docs/README.md:168]: It's not clear if this means all fields must cast successfully or not, eg. "a,14,4,b" will be `'a', 14, 4, 'b'` or `'a','14','4','b'`

Comment (tmo-trustpilot) [reply docs/README.md:168]: Also does this mean you will get different types to the handler depending on what args are passed?

**Review by tmo-trustpilot — CHANGES_REQUESTED** [review]

File: lambdarest/__init__.py (hunk 1 above)
Comment (sloev) [reply lambdarest/__init__.py:89]: 🐍 🐍 🐍 🐌  😄

**Review by tmo-trustpilot — APPROVED** [review]

**Pull request merged**